pubsub/google/cloud/pubsub_v1/publisher/batch.py

```diff
@@ -0,0 +1,231 @@
+        # number of futures I have, then something went wrong.
+        if len(response.message_ids) != len(self._.futures):
+            raise exceptions.PublishError(
+                'Some messages were not successfully published.',
+            )
+
+        # Iterate over the futures on the queue and return the response IDs.
+        # We are trusting that there is a 1:1 mapping, and raise an exception
+        # if not.
+        self._.status = 'success'
+        for message_id, fut in zip(response.message_ids, self._.futures):
+            self._.message_ids[hash(fut)] = message_id
+            fut._trigger()
+
+    def monitor(self):
+        """Commit this batch after sufficient time has elapsed.
+
+        This simply sleeps for ``self._settings.max_latency`` seconds,
+        and then calls commit unless the batch has already been committed.
+        """
+        # Note: This thread blocks; it is up to the calling code to call it
+        # in a separate thread.
+        #
+        # Sleep for however long we should be waiting.
+        time.sleep(self._.settings.max_latency)
+
+        # If, in the intervening period, the batch started to be committed,
+        # then no-op at this point.
+        if self._.status != 'accepting messages':
+            return
+
+        # Commit.
+        return self.commit()
+
+    def publish(self, data, **attrs):
+        """Publish a single message.
+
+        .. note::
+            Messages in Pub/Sub are blobs of bytes. They are *binary* data,
+            not text. You must send data as a bytestring
+            (``bytes`` in Python 3; ``str`` in Python 2), and this library
+            will raise an exception if you send a text string.
+
+            The reason that this is so important (and why we do not try to
+            coerce for you) is because Pub/Sub is also platform independent
+            and there is no way to know how to decode messages properly on
+            the other side; therefore, encoding and decoding is a required
+            exercise for the developer.
+
+        Add the given message to this object; this will cause it to be
+        published once the batch either has enough messages or a sufficient
+        period of time has elapsed.
+
+        Args:
+            data (bytes): A bytestring representing the message body. This
+                must be a bytestring (a text string will raise TypeError).
+            attrs (Mapping[str, str]): A dictionary of attributes to be
+                sent as metadata. (These may be text strings or byte strings.)
+
+        Raises:
+            TypeError: If the ``data`` sent is not a bytestring, or if the
+                ``attrs`` are not either a ``str`` or ``bytes``.
+
+        Returns:
+            Future: An object conforming to the ``concurrent.futures.Future``
+                interface.
+        """
+        # Sanity check: Is the data being sent as a bytestring?
```

Comment:
I assert that the current approach is easier for the maintainers. There should not be a backdoor method that can take invalid data; the current approach validates as the final item before sending.